Modular block request handler

rahulksnv opened the pull request:
Submit the outstanding PRs from the old repos(these were already reviewed and approved before the repo rorg, but not yet submitted): 
Main PR: https://github.com/paritytech/substrate/pull/14014
Companion PRs: https://github.com/paritytech/polkadot/pull/7134, https://github.com/paritytech/cumulus/pull/2489

The changes in the PR:
1. ChainSync currently calls into the block request handler directly. Instead, move the block request handler behind a trait. This allows new protocols to be plugged into ChainSync.
2. BuildNetworkParams is changed so that custom relay protocol implementations can be (optionally) passed in during network creation time. If custom protocol is not specified, it defaults to the existing block handler
3. BlockServer and BlockDownloader traits are introduced for the protocol implementation. The existing block handler has been changed to implement these traits
4. Other changes:
[X] Make TxHash serializable. This is needed for exchanging the serialized hash in the relay protocol messages
[X] Clean up types no longer used(OpaqueBlockRequest, OpaqueBlockResponse)

altonen commented:
I haven't looked at the changes yet but this may have to wait a little while since @dmitry-markin is doing refactoring on `ChainSync`, especially w.r.t. to request handling.

nazar-pc commented:
PR was out at the old repo since April, I was really hoping we can upstream this since it touches pretty fundamental internals and every time something changes upstream it is a bit of PITA to rebase it.

altonen commented:
Understandable and apologies for stalling with the review of the original PR. We've finally found time to start refactoring the syncing code which is long overdue and the idea is to convert `ChainSync` to do no I/O or polling, which, after a brief look, is at odds with the changes of this PR. We're most likely doing a facelift for the request/response code relatively soon as well.

@dmitry-markin can you give an estimate how much work is it to rebase your changes on top of these changes?

dmitry-markin commented on substrate/client/network/sync/src/lib.rs:
```diff
@@ -1915,22 +1818,34 @@ where
 								debug!(
-									target: LOG_TARGET,
-									"Failed to decode block response from peer {id:?}: {e:?}.",
+									target: "sync",
```
```suggestion
									target: LOG_TARGET,
```

dmitry-markin commented on substrate/client/network/sync/src/lib.rs:
```diff
@@ -1915,22 +1818,34 @@ where
-							return Poll::Ready(import)
+							Err(BlockResponseError::ExtractionFailed(e)) => {
+								debug!(
+									target: "sync",
```
```suggestion
									target: LOG_TARGET,
```

dmitry-markin commented on substrate/client/network/sync/src/block_relay_protocol.rs:
```diff
@@ -0,0 +1,72 @@
+/// that can be used to initiate concurrent downloads.
+#[async_trait::async_trait]
+pub trait BlockDownloader<Block: BlockT>: Send + Sync {
+	/// Performs the protocol specific sequence to fetch the block from the peer.
+	/// Output: if the download succeeds, the response is a `Vec<u8>` which is
+	/// in a format specific to the protocol implementation. The block data
+	/// can be extracted from this response using [`BlockDownloader::block_response_into_blocks`].
+	async fn download_block(
```
This is kind of a nitpick, but could you rename this function into `download_blocks`, please, to correctly reflect what it's doing?
```suggestion
	/// Performs the protocol specific sequence to fetch the blocks from the peer.
	/// Output: if the download succeeds, the response is a `Vec<u8>` which is
	/// in a format specific to the protocol implementation. The block data
	/// can be extracted from this response using [`BlockDownloader::block_response_into_blocks`].
	async fn download_blocks(
```

dmitry-markin commented:
> @dmitry-markin can you give an estimate how much work is it to rebase your changes on top of these changes?

I think we can merge this, the changes of this PR are mostly self contained and don't interfere in a bad way (and even simplify some things) with work on extracting request-responses. I'll just move the `BlockDownloader` one level up to the `SyncingEngine`.

As a side note, the situation around this PR seems quite weird, because the original PR was approved a long time ago, just nobody pressed the "merge" button.

altonen commented on substrate/client/network/sync/src/lib.rs:
```diff
@@ -1248,30 +1184,18 @@ where
-				);
-			},
+			let downloader = self.block_downloader.clone();
+			self.pending_responses.insert(
+				who,
+				Box::pin(async move {
+					(
+						who,
+						PeerRequest::Block(request.clone()),
+						downloader.download_block(who, request).await,
+					)
+				}),
+			);
```
This is something I would like to get refactored in the future. It's not related to this PR but you should know since you're working on your own syncing implementation.

I'd prefer having something like this:

```rust
enum RequestResponseError {
    Timeout
    Rejected
}

enum RequestResponseEvent {
    RequestReceived {
        peer: PeerId,
        request: Vec<u8>
    },
    ResponseReceived {
        peer: PeerId,
        response: Vec<u8>
    },
    RequestFailed {
        peer: PeerId,
        error: RequestResponseError,
    }
}

impl RequestResponseHandle {
    async fn send_request(peer: PeerId, request: Vec<u8>) {}
    async fn send_response(peer: PeerId, response: Vec<u8>) {}
    fn reject_request(peer: PeerId) {}
    async fn next_event(&mut self) -> Option<RequestResponseEvent> {}
}
```

Each reqeuest/response protocol would have its own handle which would allow sending requests/responses and would also allow polling both using the same object. This would eliminate using `FuturesUnordered` for polling responses. This would mean that both request and response processing is moved to `SyncingEngine`.

dmitry-markin replied:
Thanks for the insights, I'll try to implement this after requests-responses are moved to `SyncingEngine`.

altonen replied:
This is not as critical as the rest of the Sync 2.0 refactoring and it's more a future nice to have task, mainly for providing a better abstraction over libp2p.

rahulksnv replied:
Thanks for the inputs, resolving as this is won't be part of this PR

rahulksnv commented:
> As a side note, the situation around this PR seems quite weird, because the original PR was approved a long time ago, just nobody pressed the "merge" button.

Apologies on that. I had a hard time getting green on all three PRs (the main one + two in the companion repos) ..  by the time I get green, something would have changed and had to resync/retry. At some point I got busy with other stuff and this went to the back burner. So glad that the companion repos have been consolidated into one repo now

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3700046

rahulksnv commented:
@dmitry-markin @altonen Addressed the comments, please let me know what else is needed in order to merge the PR, thanks.

dmitry-markin commented:
bot fmt

nazar-pc commented:
Thanks you for quick progress here!

dmitry-markin commented:
> @dmitry-markin @altonen Addressed the comments, please let me know what else is needed in order to merge the PR, thanks.

All is good, the PR is merged. Thanks for contributing!

rahulksnv commented:
> > @dmitry-markin @altonen Addressed the comments, please let me know what else is needed in order to merge the PR, thanks.
> 
> All is good, the PR is merged. Thanks for contributing!

Thanks @dmitry-markin @altonen for the help in getting this merged

Polkadot-Forum commented:
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-release-analysis-v1-2-0/4451/1
